File: BibliotecaUFABC/app/src/main/java/com/nintersoft/bibliotecaufabc/MainActivity.java
package com.nintersoft.bibliotecaufabc;

import android.content.DialogInterface;
import android.content.Intent;
import android.content.SharedPreferences;
import android.net.Uri;
import android.os.Build;
import android.os.Bundle;

import android.preference.PreferenceManager;
import android.provider.Settings;
import android.text.InputType;
import android.util.Log;
import android.view.View;

import com.bumptech.glide.Glide;
import com.google.android.material.floatingactionbutton.FloatingActionButton;
import com.google.android.material.navigation.NavigationView;
import com.google.android.material.snackbar.Snackbar;
import com.nintersoft.bibliotecaufabc.book_search_model.BookSearchDAO;
import com.nintersoft.bibliotecaufabc.book_search_model.BookSearchDatabaseSingletonFactory;
import com.nintersoft.bibliotecaufabc.book_search_model.BookSearchProperties;
import com.nintersoft.bibliotecaufabc.synchronization.SyncService;
import com.nintersoft.bibliotecaufabc.utilities.GlobalConstants;
import com.nintersoft.bibliotecaufabc.utilities.GlobalFunctions;
import com.nintersoft.bibliotecaufabc.utilities.GlobalVariables;
import com.nintersoft.bibliotecaufabc.viewadapter.SearchBookAdapter;
import com.nintersoft.bibliotecaufabc.webviewclients.MainWebClient;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.AlertDialog;
import androidx.core.content.ContextCompat;
import androidx.core.view.GravityCompat;
import androidx.drawerlayout.widget.DrawerLayout;
import androidx.appcompat.app.ActionBarDrawerToggle;
import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.Toolbar;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import android.view.Menu;
import android.view.MenuItem;
import android.view.ViewGroup;
import android.webkit.WebView;
import android.widget.EditText;
import android.widget.LinearLayout;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.util.ArrayList;

public class MainActivity extends AppCompatActivity
        implements NavigationView.OnNavigationItemSelectedListener {

    enum PermReqState{
        BEFORE_REQ,
        REQUESTING,
        REQUESTING_LOGIN,
        REQUESTING_SYNC,
        REQUESTED
    }

    private Menu navViewMenu;
    private WebView dataSource;
    private AlertDialog loading_alert;
    private RecyclerView list;
    private MainWebClient dataClient;
    private FloatingActionButton fab;

    private BookSearchDAO dao;
    private SearchBookAdapter adapter;
    private ArrayList<BookSearchProperties> availableBooks;

    private PermReqState permReqState;

    private boolean isFirstRequest;
    private boolean hasRequestedSync;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        bindComponents();
        GlobalFunctions.createSyncNotificationChannel(getApplicationContext());
        GlobalFunctions.createRenewalNotificationChannel(getApplicationContext());
        loadPreferences();
        setWebViewSettings();
        setupBookList();
        setListeners();
    }

    @Override
    protected void onResume() {
        super.onResume();

        if (GlobalVariables.bookViewerUserNameSet != null){
            GlobalVariables.bookViewerUserNameSet = null;
            navViewMenu.findItem(R.id.nav_logout).setTitle(getString(R.string.nav_menu_logout,
                                                                        GlobalVariables.bookViewerUserNameSet));
            setUserConnected(true);

            dataClient.resetCounters();
            dataSource.loadUrl(GlobalConstants.URL_LIBRARY_NEWEST);
        }
    }

    private void bindComponents(){
        dao = BookSearchDatabaseSingletonFactory.getInstance().bookSearchDAO();

        availableBooks = new ArrayList<>();

        fab = findViewById(R.id.fab);

        Toolbar toolbar = findViewById(R.id.toolbar);
        setSupportActionBar(toolbar);

        DrawerLayout drawer = findViewById(R.id.drawer_layout);
        ActionBarDrawerToggle toggle = new ActionBarDrawerToggle(
                this, drawer, toolbar, R.string.navigation_drawer_open, R.string.navigation_drawer_close);
        drawer.addDrawerListener(toggle);
        toggle.syncState();

        NavigationView navigationView = findViewById(R.id.nav_view);
        navigationView.setNavigationItemSelectedListener(this);

        navViewMenu = navigationView.getMenu();
        navViewMenu.findItem(R.id.nav_login).setVisible(false);
        navViewMenu.findItem(R.id.nav_logout).setVisible(false);
        navViewMenu.findItem(R.id.nav_renew).setVisible(false);
        navViewMenu.findItem(R.id.nav_loans).setVisible(false);
        navViewMenu.findItem(R.id.nav_reservation).setVisible(false);

        list = findViewById(R.id.list_items_home);
        permReqState = PermReqState.BEFORE_REQ;
    }

    private void loadPreferences(){
        SharedPreferences pref = PreferenceManager.getDefaultSharedPreferences(this);
        GlobalVariables.showShare = pref.getBoolean(getString(R.string.key_general_share_app_enabled), true);
        GlobalVariables.keepCache = pref.getBoolean(getString(R.string.key_privacy_cache_main_content), true);
        GlobalVariables.ringAlarm = pref.getBoolean(getString(R.string.key_notification_enable_warning), true);
        GlobalVariables.showExtWarning = pref.getBoolean(getString(R.string.key_general_leave_warning), true);
        GlobalVariables.storeUserFormData = pref.getBoolean(getString(R.string.key_privacy_store_password), true);
        GlobalVariables.loginAutomatically = pref.getBoolean(getString(R.string.key_privacy_auto_login), true);

        String syncInterval = pref.getString(getString(R.string.key_notification_sync_interval), "2");
        GlobalVariables.syncInterval = Integer.parseInt(syncInterval == null ? "2" : syncInterval);

        String warningDelay = pref.getString(getString(R.string.key_notification_warning_delay), "0");
        GlobalVariables.ringAlarmOffset = Integer.parseInt(warningDelay == null ? "0" : warningDelay);

        isFirstRequest = true;
        hasRequestedSync = false;

        applyPreferences(pref);
    }

    private void applyPreferences(@Nullable SharedPreferences preferences){
        navViewMenu.findItem(R.id.nav_share).setVisible(GlobalVariables.showShare);

        if (!GlobalVariables.storeUserFormData && preferences != null){
            SharedPreferences.Editor editor = preferences.edit();
            editor.remove(getString(R.string.key_privacy_login_username));
            editor.remove(getString(R.string.key_privacy_login_password));
            editor.apply();
        }
    }

    private void setupBookList(){
        adapter = new SearchBookAdapter(this, availableBooks);
        list.setLayoutManager(new LinearLayoutManager(this));
        list.setAdapter(adapter);
        adapter.notifyDataSetChanged();
    }

    private void setWebViewSettings(){
        dataSource = new WebView(this);
        dataClient = new MainWebClient(this);
        GlobalFunctions.configureStandardWebView(dataSource);

        dataSource.setWebViewClient(dataClient);
        dataSource.loadUrl(GlobalConstants.URL_LIBRARY_NEWEST);

        AlertDialog.Builder builder = new AlertDialog.Builder(this);
        builder.setView(R.layout.message_progress_dialog);
        builder.setOnCancelListener(new DialogInterface.OnCancelListener() {
            @Override
            public void onCancel(DialogInterface dialog) {
                Snackbar.make(findViewById(R.id.list_items_home), R.string.snack_message_loading_newest,
                        Snackbar.LENGTH_INDEFINITE).show();
            }
        });
        loading_alert = builder.create();

        loading_alert.setOnDismissListener(new DialogInterface.OnDismissListener() {
            @Override
            public void onDismiss(DialogInterface dialog) {
                requestSyncPermission();
            }
        });
        loading_alert.setOnCancelListener(new DialogInterface.OnCancelListener() {
            @Override
            public void onCancel(DialogInterface dialog) {
                requestSyncPermission();
            }
        });
        loading_alert.show();
    }

    private void setListeners(){
        fab.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                openSearchDialog();
            }
        });
    }

    @Override
    public void onBackPressed() {
        DrawerLayout drawer = findViewById(R.id.drawer_layout);
        if (drawer.isDrawerOpen(GravityCompat.START)) {
            drawer.closeDrawer(GravityCompat.START);
        } else {
            super.onBackPressed();
        }
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        // Inflate the menu; this adds items to the action bar if it is present.
        getMenuInflater().inflate(R.menu.main, menu);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()){
            case R.id.action_settings:
                Intent acIntent = new Intent(this, SettingsActivity.class);
                startActivityForResult(acIntent, GlobalConstants.ACTIVITY_SETTINGS_REQUEST_CODE);
                return true;
            case R.id.action_refresh:
                dataClient.resetCounters();
                dataSource.loadUrl(GlobalConstants.URL_LIBRARY_NEWEST);
                Snackbar.make(findViewById(R.id.list_items_home),
                        R.string.snack_message_refreshing_newest, Snackbar.LENGTH_LONG).show();
                return true;
        }
        return super.onOptionsItemSelected(item);
    }

    @Override
    public boolean onNavigationItemSelected(@NonNull  MenuItem item) {
        switch (item.getItemId()){
            case R.id.nav_search: openSearchDialog(); break;
            case R.id.nav_renew: {
                Intent acIntent = new Intent(this, RenewalActivity.class);
                startActivityForResult(acIntent, GlobalConstants.ACTIVITY_RENEWAL_REQUEST_CODE);
                break;
            }
            case R.id.nav_loans: {
                Intent acIntent = new Intent(this, BookLoansActivity.class);
                startActivityForResult(acIntent, GlobalConstants.ACTIVITY_SETTINGS_REQUEST_CODE);
                break;
            }
            case R.id.nav_manage: {
                Intent acIntent = new Intent(this, SettingsActivity.class);
                startActivityForResult(acIntent, GlobalConstants.ACTIVITY_SETTINGS_REQUEST_CODE);
                break;
            }
            case R.id.nav_reservation: {
                Intent acIntent = new Intent(this, ReservationActivity.class);
                startActivity(acIntent);
                break;
            }
            case R.id.nav_login: {
                Intent acIntent = new Intent(this, LoginActivity.class);
                startActivityForResult(acIntent, GlobalConstants.ACTIVITY_LOGIN_REQUEST_CODE);
                break;
            }
            case R.id.nav_logout:
                dataSource.loadUrl(GlobalConstants.URL_LIBRARY_LOGOUT);
                navViewMenu.findItem(R.id.nav_login).setVisible(false);
                navViewMenu.findItem(R.id.nav_logout).setVisible(false);
                break;
            case R.id.nav_share: {
                Intent share = new Intent(Intent.ACTION_SEND);
                share.setType("text/plain");
                share.putExtra(Intent.EXTRA_TEXT, getString(R.string.share_app_structure));
                if (share.resolveActivity(getPackageManager()) != null)
                    startActivity(Intent.createChooser(share, getString(R.string.intent_share_app)));
                break;
            }
            case R.id.nav_quit: finish(); break;
        }

        DrawerLayout drawer = findViewById(R.id.drawer_layout);
        drawer.closeDrawer(GravityCompat.START);
        return true;
    }

    @Override
    protected void onDestroy() {
        if (!GlobalVariables.keepCache){
            // Clear Glide cache
            Glide.get(getApplicationContext()).clearDiskCache();

            // Clear WebView cache
            dataSource.clearCache(true);
            dataSource.clearFormData();
            dataSource.clearHistory();
        }

        super.onDestroy();
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, @Nullable Intent data) {
        super.onActivityResult(requestCode, resultCode, data);

        if (requestCode == GlobalConstants.ACTIVITY_LOGIN_REQUEST_CODE){
            if (resultCode == RESULT_OK  && data != null){
                String userName = data.getStringExtra(GlobalConstants.CONNECTED_STATUS_USER_NAME);
                navViewMenu.findItem(R.id.nav_logout).setTitle(getString(R.string.nav_menu_logout, userName == null ? "???" : userName));
                setUserConnected(true);

                dataClient.resetCounters();
                dataSource.loadUrl(GlobalConstants.URL_LIBRARY_NEWEST);
            }
            if (!hasRequestedSync){
                GlobalFunctions.scheduleNextSynchronization(this, GlobalFunctions.nextStandardSync());
                ContextCompat.startForegroundService(this,
                        new Intent(this, SyncService.class).putExtra("service", false));
                hasRequestedSync = true;
            }
        }
        else if (requestCode == GlobalConstants.ACTIVITY_SETTINGS_REQUEST_CODE)
            loadPreferences();
        else if (requestCode == GlobalConstants.SYNC_PERMISSION_REQUEST_ID){
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
                if (!Settings.canDrawOverlays(this)) {
                    AlertDialog.Builder builder = new AlertDialog.Builder(this);
                    builder.setTitle(R.string.notification_sync_denied_title)
                            .setMessage(R.string.notification_sync_denied_message)
                            .setPositiveButton(R.string.dialog_button_ok, null)
                            .create().show();
                }

                if (permReqState == PermReqState.REQUESTING_LOGIN){
                    permReqState = PermReqState.REQUESTED;
                    if (GlobalVariables.loginAutomatically){
                        Intent acIntent = new Intent(this, LoginActivity.class);
                        startActivityForResult(acIntent, GlobalConstants.ACTIVITY_LOGIN_REQUEST_CODE);
                    }
                    else {
                        GlobalFunctions.scheduleNextSynchronization(this, GlobalFunctions.nextStandardSync());
                        ContextCompat.startForegroundService(this,
                                new Intent(this, SyncService.class).putExtra("service", false));
                        hasRequestedSync = true;
                    }
                    isFirstRequest = false;
                }
                else {
                    GlobalFunctions.scheduleNextSynchronization(this, GlobalFunctions.nextStandardSync());
                    ContextCompat.startForegroundService(this,
                            new Intent(this, SyncService.class).putExtra("service", false));
                    hasRequestedSync = true;
                }
            }
        }
    }

    private void openSearchDialog(){
        final Intent acIntent = new Intent(this, SearchActivity.class);

        final EditText input = new EditText(this);
        input.setInputType(InputType.TYPE_TEXT_FLAG_AUTO_COMPLETE | InputType.TYPE_TEXT_FLAG_AUTO_CORRECT);

        LinearLayout layout = new LinearLayout(this);
        LinearLayout.LayoutParams params = new LinearLayout.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT,
                ViewGroup.LayoutParams.WRAP_CONTENT);
        params.setMargins(getResources().getDimensionPixelSize(R.dimen.dialog_view_margin), 0,
                getResources().getDimensionPixelSize(R.dimen.dialog_view_margin), 0);
        input.setLayoutParams(params);
        layout.addView(input);

        AlertDialog.Builder alertBuilder = new AlertDialog.Builder(this);
        alertBuilder.setTitle(R.string.dialog_input_title_search);
        alertBuilder.setView(layout);
        alertBuilder.setPositiveButton(R.string.dialog_button_search, new DialogInterface.OnClickListener() {
            @Override
            public void onClick(DialogInterface dialog, int which) {
                String terms = input.getText().toString();
                if (terms.isEmpty()) return;

                acIntent.putExtra("terms", input.getText().toString());
                startActivityForResult(acIntent, GlobalConstants.ACTIVITY_SEARCH_REQUEST_CODE);
                dialog.dismiss();
            }
        });
        alertBuilder.create().show();
    }

    public void setSearchResults(String results){
        if (results.isEmpty()) return;

        try {
            dao.removeAll();
            availableBooks.clear();
            JSONArray jsResultsArr = new JSONArray(results);
            for (int i = 0; i < jsResultsArr.length(); ++i){
                JSONObject jsBook = jsResultsArr.getJSONObject(i);
                BookSearchProperties newBook = new BookSearchProperties();
                newBook.setTitle(jsBook.getString("title"));
                newBook.setAuthor(jsBook.getString("author"));
                newBook.setSection(jsBook.getString("section"));
                newBook.setType(jsBook.getString("type"));
                newBook.setCode(jsBook.getString("code"));
                newBook.setId(i);
                availableBooks.add(newBook);
                dao.insert(newBook);
            }
            adapter.notifyDataSetChanged();
        } catch (JSONException e) {
            Snackbar.make(findViewById(R.id.list_items_home),
                    R.string.snack_message_parse_fail, Snackbar.LENGTH_LONG)
                    .show();
        } finally {
            navViewMenu.findItem(R.id.nav_loans).setVisible(false);
            if (loading_alert.isShowing()) loading_alert.dismiss();
            else Snackbar.make(findViewById(R.id.list_items_home), R.string.snack_message_loaded_newest,
                    Snackbar.LENGTH_SHORT).show();
        }
    }

    public void setUserConnected(boolean connected, String userName){
        navViewMenu.findItem(R.id.nav_logout).setTitle(getString(R.string.nav_menu_logout, userName == null ? "???" : userName));
        setUserConnected(connected);
    }

    synchronized private void setUserConnected(boolean connected){
        navViewMenu.findItem(R.id.nav_login).setVisible(!connected);
        navViewMenu.findItem(R.id.nav_logout).setVisible(connected);
        navViewMenu.findItem(R.id.nav_renew).setVisible(connected);
        navViewMenu.findItem(R.id.nav_reservation).setVisible(connected);

        if (isFirstRequest){
            if (permReqState == PermReqState.REQUESTING || permReqState == PermReqState.BEFORE_REQ){
                permReqState = connected ? PermReqState.REQUESTING_SYNC : PermReqState.REQUESTING_LOGIN;
                return;
            }

            if (!connected && GlobalVariables.loginAutomatically){
                Intent acIntent = new Intent(this, LoginActivity.class);
                startActivityForResult(acIntent, GlobalConstants.ACTIVITY_LOGIN_REQUEST_CODE);
            }
            else {
                GlobalFunctions.scheduleNextSynchronization(this, GlobalFunctions.nextStandardSync());
                ContextCompat.startForegroundService(this,
                        new Intent(this, SyncService.class).putExtra("service", false));
                hasRequestedSync = true;
            }
            isFirstRequest = false;
        }
    }

    public void receiveError(String message){
        if (loading_alert.isShowing()) loading_alert.dismiss();

        Snackbar snack_error = Snackbar.make(findViewById(R.id.list_items_home),
                getString(R.string.snack_message_loading_main_error, message), Snackbar.LENGTH_INDEFINITE);
        snack_error.setAction(R.string.snack_button_reload,
                new View.OnClickListener() {
                    @Override
                    public void onClick(View v) {
                        dataClient.resetCounters();
                        dataSource.loadUrl(GlobalConstants.URL_LIBRARY_NEWEST);
                        Snackbar.make(findViewById(R.id.list_items_home),
                                R.string.snack_message_loading_newest, Snackbar.LENGTH_LONG).show();
                    }
                });
        snack_error.show();

        availableBooks.clear();
        availableBooks.addAll(dao.getAll());
        adapter.notifyDataSetChanged();

        if (GlobalVariables.ringAlarm) navViewMenu.findItem(R.id.nav_loans).setVisible(true);
    }

    synchronized public void requestSyncPermission(){
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M && !Settings.canDrawOverlays(this)) {
            permReqState = permReqState == PermReqState.BEFORE_REQ ? PermReqState.REQUESTING : PermReqState.REQUESTING_LOGIN;

            AlertDialog.Builder builder = new AlertDialog.Builder(this);
            builder.setTitle(R.string.notification_sync_rationale_title)
                    .setMessage(R.string.notification_sync_rationale_message)
                    .setPositiveButton(R.string.dialog_button_ok, new DialogInterface.OnClickListener() {
                        @Override
                        public void onClick(DialogInterface dialog, int which) {
                            dialog.cancel();
                            Intent intent = new Intent(Settings.ACTION_MANAGE_OVERLAY_PERMISSION,
                                    Uri.parse("package:" + getPackageName()));
                            if (intent.resolveActivity(getPackageManager()) != null)
                                startActivityForResult(intent, GlobalConstants.SYNC_PERMISSION_REQUEST_ID);
                        }
                    }).create().show();
        }
        else if (permReqState == PermReqState.REQUESTING_LOGIN) {
            Intent acIntent = new Intent(this, LoginActivity.class);
            startActivityForResult(acIntent, GlobalConstants.ACTIVITY_LOGIN_REQUEST_CODE);
            permReqState = PermReqState.REQUESTED;
        }
        else if (permReqState == PermReqState.REQUESTING_SYNC){
            GlobalFunctions.scheduleNextSynchronization(this, GlobalFunctions.nextStandardSync());
            ContextCompat.startForegroundService(this,
                    new Intent(this, SyncService.class).putExtra("service", false));
            hasRequestedSync = true;
        }
        else permReqState = PermReqState.REQUESTED;
    }
}

File: BibliotecaUFABC/app/src/main/java/com/nintersoft/bibliotecaufabc/activities/MainActivity.kt
package com.nintersoft.bibliotecaufabc.activities

import android.app.Activity
import android.content.Intent
import android.content.SharedPreferences
import android.net.Uri
import android.os.Build
import android.os.Bundle
import android.provider.Settings
import android.webkit.CookieManager
import android.webkit.WebView
import androidx.appcompat.app.AlertDialog
import com.google.android.material.bottomnavigation.BottomNavigationView
import androidx.appcompat.app.AppCompatActivity
import androidx.core.content.ContextCompat
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProvider
import androidx.navigation.findNavController
import androidx.navigation.ui.AppBarConfiguration
import androidx.navigation.ui.setupActionBarWithNavController
import androidx.navigation.ui.setupWithNavController
import androidx.preference.PreferenceManager
import androidx.work.PeriodicWorkRequest
import androidx.work.WorkInfo
import androidx.work.WorkManager
import com.bumptech.glide.Glide
import com.google.android.material.snackbar.Snackbar
import com.nintersoft.bibliotecaufabc.R
import com.nintersoft.bibliotecaufabc.global.Constants
import com.nintersoft.bibliotecaufabc.global.Functions
import com.nintersoft.bibliotecaufabc.synchronization.SyncService
import com.nintersoft.bibliotecaufabc.ui.home.HomeFragment
import com.nintersoft.bibliotecaufabc.ui.home.HomeViewModel
import com.nintersoft.bibliotecaufabc.ui.reservations.ReservationsRecyclerFragment
import com.nintersoft.bibliotecaufabc.ui.reservations.ReservationsViewModel
import com.nintersoft.bibliotecaufabc.ui.snackbar.MessageSnackbar
import com.nintersoft.bibliotecaufabc.webclient.HomeWebClient
import com.nintersoft.bibliotecaufabc.webclient.ReservationWebClient
import kotlinx.android.synthetic.main.activity_main.*
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.launch
import java.lang.Exception
import java.util.*
import java.util.concurrent.ExecutionException
import java.util.concurrent.TimeUnit

class MainActivity : AppCompatActivity(), ReservationsRecyclerFragment.ReservationEvents,
    HomeFragment.HomeFragmentListener {

    private var dataSource : WebView? = null
    private var reservationSource : WebView? = null

    private lateinit var prefs : SharedPreferences
    private lateinit var loadingAlert : AlertDialog
    private lateinit var homeViewModel : HomeViewModel
    private lateinit var reservationsViewModel : ReservationsViewModel

    private val prefListener = SharedPreferences.OnSharedPreferenceChangeListener { _, key ->
        if(key == getString(R.string.key_general_share_app_enabled)) invalidateOptionsMenu()
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        val navView: BottomNavigationView = findViewById(R.id.nav_view)
        prefs = PreferenceManager.getDefaultSharedPreferences(this)
        homeViewModel = ViewModelProvider(this)[HomeViewModel::class.java]
        reservationsViewModel = ViewModelProvider(this)[ReservationsViewModel::class.java]

        val navController = findNavController(R.id.nav_host_fragment)
        val appBarConfiguration = AppBarConfiguration(
            setOf(
                R.id.navigation_home,
                R.id.navigation_renewal,
                R.id.navigation_reservation
            )
        )
        setupActionBarWithNavController(navController, appBarConfiguration)
        navView.setupWithNavController(navController)

        Functions.createNotificationChannel(
            getString(R.string.notification_sync_channel_description),
            getString(R.string.notification_sync_channel_description),
            Constants.CHANNEL_SYNC_ID
        )
        Functions.createNotificationChannel(
            getString(R.string.notification_renewal_channel_title),
            getString(R.string.notification_renewal_channel_description),
            Constants.CHANNEL_RENEWAL_ID
        )

        configureWebView()
        configureObservers()
        prefs.registerOnSharedPreferenceChangeListener(prefListener)
    }

    override fun onDestroy() {
        prefs.unregisterOnSharedPreferenceChangeListener(prefListener)
        if (!prefs.getBoolean(getString(R.string.key_privacy_cache_main_content), true)){
            GlobalScope.launch { Glide.get(applicationContext).clearDiskCache() }
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
                CookieManager.getInstance().removeAllCookies(null)
                CookieManager.getInstance().flush()
            }
            dataSource?.clearCache(true)
            dataSource?.clearFormData()
            dataSource?.clearHistory()
        }
        super.onDestroy()
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        when (requestCode){
            Constants.ACTIVITY_LOGIN_REQUEST_CODE -> {
                if (resultCode == Activity.RESULT_OK){
                    val user = data?.getStringExtra(Constants.CONNECTED_STATUS_USER_NAME)
                    if (!user.isNullOrEmpty()){
                        homeViewModel.defineLoginStatus(true)
                        homeViewModel.defineConnectedUserName(user)
                    }
                }
            }
            Constants.SYNC_PERMISSION_REQUEST_ID -> {
                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M){
                    if (!Settings.canDrawOverlays(this))
                        AlertDialog.Builder(this).apply {
                            setTitle(R.string.notification_sync_denied_title)
                            setMessage(R.string.notification_sync_denied_message)
                            setPositiveButton(R.string.dialog_button_ok, null)
                        }.create().show()
                    else {
                        startPowerSavingSettings()
                        setSyncSchedule()
                    }
                }
            }
        }
    }

    override fun onRequestRefresh() {
        dataSource?.loadUrl(Constants.URL_LIBRARY_NEWEST)
        MessageSnackbar.make(container, R.string.snack_message_refreshing_newest,
            Snackbar.LENGTH_SHORT)?.setAnchorView(nav_view)?.show()
    }

    override fun onRequestSignOut() {
        dataSource?.loadUrl(Constants.URL_LIBRARY_LOGOUT)
        homeViewModel.askedLogout.value = true
        homeViewModel.defineLoginStatus(false)
        homeViewModel.defineConnectedUserName("")
        invalidateOptionsMenu()
    }

    override fun loadCancelLink(link: String) { reservationSource?.loadUrl(link) }

    override fun refreshReservations() { reservationSource?.
        loadUrl(Constants.URL_LIBRARY_RESERVATION) }

    private fun configureWebView(){
        dataSource = WebView(this)
        reservationSource = WebView(this)
        Functions.configureWebView(dataSource!!, HomeWebClient(homeViewModel))
        Functions.configureWebView(reservationSource!!,
            ReservationWebClient(reservationsViewModel))

        loadingAlert = AlertDialog.Builder(this).apply {
            setView(R.layout.message_progress_dialog)
            setOnCancelListener {
                MessageSnackbar.make(container, R.string.snack_message_loading_newest,
                    Snackbar.LENGTH_SHORT)?.setAnchorView(nav_view)?.show()
            }
        }.create()
        loadingAlert.show()
    }

    private fun configureObservers(){
        homeViewModel.loginStatus.observe(this, {
            when (it){
                null -> dataSource?.loadUrl(Constants.URL_LIBRARY_NEWEST)
                true -> {
                    if (loadingAlert.isShowing) loadingAlert.dismiss()
                    reservationSource?.loadUrl(Constants.URL_LIBRARY_RESERVATION)
                    homeViewModel.askedLogout.value = false
                    homeViewModel.hasCheckedPermission.let {has ->
                        if (!has.value!!) {
                            requestSyncPermission()
                            has.value = true
                        }
                    }
                }
                else -> {
                    if (loadingAlert.isShowing) loadingAlert.dismiss()
                    if (prefs.getBoolean(getString(R.string.key_privacy_auto_login), true) &&
                            !homeViewModel.hasRequestedLogin.value!!) {
                        startActivityForResult(
                            Intent(this, LoginActivity::class.java),
                            Constants.ACTIVITY_LOGIN_REQUEST_CODE
                        )
                        homeViewModel.hasRequestedLogin.value = true
                    }
                }
            }
            nav_view.menu.findItem(R.id.navigation_reservation).isEnabled = it == true
        })

        homeViewModel.dataLoadError.observe(this, Observer {
            if (it == null) return@Observer
            if (loadingAlert.isShowing) loadingAlert.dismiss()
            MessageSnackbar.make(container, getString(R.string.snack_message_loading_main_error),
                Snackbar.LENGTH_LONG, MessageSnackbar.Type.ERROR)?.setAnchorView(nav_view)?.apply {
                setAction(R.string.snack_button_reload, ({
                    dataSource?.loadUrl(Constants.URL_LIBRARY_NEWEST)
                    Snackbar.make(container, R.string.snack_message_loading_newest,
                        Snackbar.LENGTH_SHORT).setAnchorView(nav_view).show()
                }))
            }?.show()
        })

        homeViewModel.connectedUserName.observe(this, Observer {
            if (it.isNullOrEmpty()) return@Observer
            homeViewModel.hasRequestedLogin.value = true
            MessageSnackbar.make(container, getString(R.string.snack_message_connected, it),
                Snackbar.LENGTH_LONG, MessageSnackbar.Type.INFO)?.setAnchorView(nav_view)?.show()
        })

        reservationsViewModel.loadError.observe(this, {
            if (it == true) MessageSnackbar.make(container,
                getString(R.string.snack_message_loading_reservation_error),
                Snackbar.LENGTH_LONG, MessageSnackbar.Type.ERROR)?.setAnchorView(nav_view)?.show()
        })

        BookViewerActivity.loggedInAs.observe(this, {
            if (!it.isNullOrEmpty()){
                homeViewModel.defineLoginStatus(true)
                homeViewModel.defineConnectedUserName(it)
            }
        })
    }

    @Synchronized
    private fun requestSyncPermission(){
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M &&
            !Settings.canDrawOverlays(this)){
            prefs.edit().putBoolean(getString(R.string.key_app_first_run), true).apply()

            AlertDialog.Builder(this).apply {
                setTitle(R.string.notification_sync_rationale_title)
                setMessage(R.string.notification_sync_rationale_message)
                setPositiveButton(R.string.dialog_button_ok, ({ dlg, _ ->
                    dlg.cancel()
                    Intent(Settings.ACTION_MANAGE_OVERLAY_PERMISSION,
                        Uri.parse("package:$packageName")).also {
                        if (it.resolveActivity(packageManager) != null)
                            startActivityForResult(it, Constants.SYNC_PERMISSION_REQUEST_ID)
                    }
                }))
            }.create().show()
        }
        else if (!homeViewModel.hasRequestedSync.value!!) {
            setSyncSchedule()

            if (SyncService.isRunning.value == false)
                ContextCompat.startForegroundService(this,
                    Intent(this, SyncService::class.java).
                        putExtra(Constants.SYNC_INTENT_SCHEDULED, false))
            homeViewModel.hasRequestedSync.value = true
        }
    }

    private fun setSyncSchedule(force : Boolean = false){
        val workState = WorkManager.getInstance(applicationContext).
            getWorkInfosForUniqueWork(Constants.WORK_SYNC_SERVICE_WORKER).let {
            try {
                if (it.get().size == 0) WorkInfo.State.CANCELLED
                else it.get()[0].state
            } catch (_ : ExecutionException) { WorkInfo.State.CANCELLED }
            catch (_ : InterruptedException) { WorkInfo.State.CANCELLED }
        }
        if (prefs.getBoolean(getString(R.string.key_app_first_run), true) || force ||
            ((workState != WorkInfo.State.RUNNING && workState != WorkInfo.State.ENQUEUED)
                    && prefs.getBoolean(getString(R.string.key_notification_enable_warning),
                        true))){
            Functions.schedulePeriodicSync(TimeUnit.DAYS.toMillis(prefs.
                getString(getString(R.string.key_notification_sync_interval), "2")!!.toLong()),
                PeriodicWorkRequest.MIN_PERIODIC_FLEX_MILLIS)
            Functions.schedulePeriodicSyncReminder(TimeUnit.DAYS.toMillis(Constants.
                SYNC_REMINDER_NOTIFICATION_INTERVAL), PeriodicWorkRequest.MIN_PERIODIC_FLEX_MILLIS,
                TimeUnit.DAYS.toMillis(1))
            prefs.edit().putBoolean(getString(R.string.key_app_first_run), false).apply()
        }
    }

    private fun startPowerSavingSettings(){
        Constants.POWER_MANAGER_INTENTS.forEach {
            if (Functions.isCallable(it)){
                AlertDialog.Builder(this).apply {
                    setTitle(getString(R.string.dialog_permission_title, Build.MANUFACTURER))
                    setMessage(getString(R.string.dialog_permission_message,
                        getString(R.string.app_name)))
                    setPositiveButton(R.string.dialog_button_yes, ({ _, _ ->
                        if (Build.MANUFACTURER.toLowerCase(Locale.getDefault()) == "xiaomi"){
                            try {
                                startActivity(Intent(Settings.ACTION_APPLICATION_DETAILS_SETTINGS).
                                    setData(Uri.parse("package:$packageName")))
                            } catch (_ : Exception){ startActivity(it) }
                        }
                        else startActivity(it)
                    }))
                    setNegativeButton(R.string.dialog_button_no, null)
                }.create().show()
                return@forEach
            }
        }
    }
}
